lib/core/src/Cardano/Pool/Metrics.hs

```diff
@@ -52,8 +73,105 @@ import GHC.Generics
 import Numeric.Natural
     ( Natural )
 
+import qualified Data.List.NonEmpty as NE
 import qualified Data.Map.Strict as Map
 
+data Block = Block
+    { header :: BlockHeader
+    , producer :: Maybe PoolId
+    } deriving (Eq, Show, Generic)
+
+data ErrMetricsWorkerErr
+    = ErrMetricsWorkerErrTipTooVolatile
+    | ErrMetricsWorkerStakeDistributionUnavailible
+    | ErrMetricsWorkerDataInconsistencyBlockWithoutProducer
+    | ErrMetricsWorkerDataInconsistencyDuplicateBlock
+    | ErrMetricsWorkerTipIsUnreachable
+    deriving (Show, Eq)
+
+-- | @worker@ follows the chain and puts pool productions and stake
+-- distributions to a @DBLayer@, such that the data in the @DB@ always is
+-- consistent.
+--
+-- The pool productions and stake distrubtions in the DB can /never/ be from
+-- different forks.
+worker
+    :: NetworkLayer IO t Block
+    -> DBLayer IO
+    -> Trace IO Text
+    -> IO ()
+worker nl db tr = do
+    -- Read the k latest headers the DB knows about
+    let (_block0, bp) = staticBlockchainParameters nl
+    let k = fromIntegral . getQuantity . view #getEpochStability $ bp
+    knownHeaders <- readCursor db k
+
+    follow nl tr knownHeaders advance' rollback getHeader
+  where
+
+    advance'
+        :: NE.NonEmpty Block
+        -> BlockHeader
+        -> IO (FollowAction ErrMetricsWorkerErr)
+    advance' bs h = runExceptT (advance bs h) >>= return . \case
+        Left ErrMetricsWorkerErrTipTooVolatile
+            -> Retry
+        Left e
+            -> ExitWith e
+            -- TODO: Make all cases explicit here.
```

Comment:
Not 100% if this indirection is good or if `advance'` should be merged into `advance`.